Sources/WebIDL/Tokenizer/Tokenizer.swift

```diff
@@ -159,164 +180,173 @@ public enum Tokenizer {
             guard let double = Double(buffer) else {
                 return
             }
-            tokens.append(.decimal)
+            appendToken(.decimal)
             decimals.append(double)
         }
 
         func appendIdentifier() {
             if let symbol = Terminal(rawValue: buffer) {
-                tokens.append(.terminal(symbol))
+                appendToken(.terminal(symbol))
             } else if identifierRegex.match(buffer) {
-                tokens.append(.identifier)
+                appendToken(.identifier)
                 identifiers.append(buffer)
             } else if integerRegex.match(buffer), let integer = Int(buffer) ?? Int(buffer, radix: 16) {
-                tokens.append(.integer)
+                appendToken(.integer)
                 integers.append(integer)
             } else if decimalRegex.match(buffer), let double = Double(buffer) {
-                tokens.append(.decimal)
+                appendToken(.decimal)
                 decimals.append(double)
             } else if otherRegex.match(buffer) {
-                tokens.append(.other)
+                appendToken(.other)
                 others.append(buffer)
             } else {
-                print("[\(name)] Undefined sequence: \(buffer)")
+                print("[\(filePath):\(currentRange())] Undefined sequence: \(buffer)")
             }
             reset()
         }
 
-        for character in string {
+        while currentIndex < string.utf8.endIndex, let character = String(
+                decoding: [string.utf8[currentIndex]],
+                as: UTF8.self
+        ).first {
+            defer {
+                if currentOffset < string.utf8.count - 1 {
+                    currentOffset += 1
+                }
+                currentIndex = string.utf8.index(after: currentIndex)
+            }
```

Comment:
How does this work with non-ASCII characters (like emoji variable names) where a single byte of the UTF-8 representation can’t be converted to a Unicode string?

Reply:
We're going through UTF-8 code units here, as [`SourceLocation` counts everything in UTF-8 byte offsets](https://github.com/apple/swift-syntax/blob/d8aa974f0532d7f0ce42b87afe476d5e4e9211ab/Sources/SwiftSyntax/SourceLocation.swift#L52) . Grapheme clusters composed of multiple code units (like some emojis) would be treated as separate symbols here, but as we don't expect them outside of comments I don't think that's a big problem if any at all.

Reply:
I found this article to be helpful in clarifying this for me https://oleb.net/blog/2017/11/swift-4-strings/